libethereum/EthereumPeer.cpp

```diff
@@ -246,22 +246,80 @@ bool EthereumPeer::interpret(unsigned _id, RLP const& _r)
 	}
-	case GetBlockHashesPacket:
+	case GetBlockHeadersPacket:
 	{
-		h256 later = _r[0].toHash<h256>();
-		unsigned limit = _r[1].toInt<unsigned>();
-		clog(NetMessageSummary) << "GetBlockHashes (" << limit << "entries," << later << ")";
-		unsigned c = min<unsigned>(host()->chain().number(later), min(limit, c_maxHashesToSend));
+		/// Packet layout:
+		/// [ block: { P , B_32 }, maxHeaders: P, skip: P, reverse: P in { 0 , 1 } ]
+		auto blockId = _r[0];
+		auto maxHeaders = _r[1].toInt<bigint>();
+		auto skip = _r[2].toInt<bigint>();
+		auto reverse = _r[3].toInt<bool>();
+
+		bigint blockNumber;
+		if (blockId.size() == 32) // block id is a hash
+		{
+			auto blockHash = blockId.toHash<h256>();
+			blockNumber = host()->chain().number(blockHash);
+			clog(NetMessageSummary) << "GetBlockHeaders (block (hash): " << blockHash
+			<< ", maxHeaders: " << maxHeaders
+			<< ", skip: " << skip << ", reverse: " << reverse << ")";
+		}
+		else // block id is a number
+		{
+			blockNumber = blockId.toInt<bigint>();
+			clog(NetMessageSummary) << "GetBlockHeaders (block (number): " << blockNumber
+			<< ", maxHeaders: " << maxHeaders
+			<< ", skip: " << skip << ", reverse: " << reverse << ")";
+		}
+
+		// Find a top requested block
+		if (reverse)
+		{
+			blockNumber -= skip; // in descending order just skip blocks
+			if (blockNumber < 0)
+			{
+				// TODO: Better solution would be to use safe u256 and treat all catches of overflows as stupid requests.
+				clog(NetImpolite) << "Requesting negative starting block.";
+				addRating(-10);
+				break;
+			}
+		}
+		else
+			blockNumber += skip + maxHeaders - 1; // in ascending order find number of the block that end the requested subset
+
+		auto lastBlockNumber = host()->chain().number();
+		if (blockNumber > lastBlockNumber)
+		{
+			maxHeaders -= (blockNumber - lastBlockNumber);
+			if (maxHeaders < 0) // we don't have any of requested blocks
+				maxHeaders = 0;
+			blockNumber = lastBlockNumber;
+		}
+
+		maxHeaders = std::min(maxHeaders, bigint{c_maxHashesToSend}); // limit the response size according to local config
+
+		// These asserts validate the logic above. In case we missed something we will send a stupid answer.
+		assert(blockNumber >= 0 && blockNumber <= lastBlockNumber && "Invalid calculated block number");
+		assert(maxHeaders >= 0 && maxHeaders <= c_maxHashesToSend && "Invalid calculated number of headers");
+
+		auto numHeadersToSend = static_cast<decltype(c_maxHashesToSend)>(maxHeaders);
+		auto hash = host()->chain().numberHash(static_cast<unsigned>(blockNumber));
+
 		RLPStream s;
-		prep(s, BlockHashesPacket, c);
-		h256 p = host()->chain().details(later).parent;
-		for (unsigned i = 0; i < c && p; ++i, p = host()->chain().details(p).parent)
-			s << p;
+		prep(s, BlockHeadersPacket, numHeadersToSend);
+		for (unsigned i = 0; i != numHeadersToSend; ++i)
+		{
+			if (!hash)
+				break;
+
+			s << host()->chain().headerData(hash);
+			hash = host()->chain().details(hash).parent;
```

Comment:
`skip` and `reverse` should not affect initial value, but how following block numbers are calculated. You  can still do enumeration by hashes skipping `skip` hashes if `reverse == 1`, although it will be inefficient for large `skip`, but you'll have to enumerate by number if `reverse == 0`